node/engine/policy/fairoutcome.go

```diff
@@ -0,0 +1,63 @@
+package policy
+
+import (
+	"log/slog"
+	"math/big"
+
+	"github.com/statechannels/go-nitro/internal/logging"
+	"github.com/statechannels/go-nitro/payments"
+	"github.com/statechannels/go-nitro/protocols"
+	"github.com/statechannels/go-nitro/protocols/directfund"
+	"github.com/statechannels/go-nitro/protocols/virtualfund"
+	"github.com/statechannels/go-nitro/types"
+)
+
+type fairOutcomePolicy struct {
+	me types.Address
+}
+
+func NewFairOutcomePolicy(me types.Address) PolicyMaker {
+	return &fairOutcomePolicy{me: me}
+}
+
+// ShouldApprove decides to approve o if it is currently unapproved
+func (fp *fairOutcomePolicy) ShouldApprove(o protocols.Objective) bool {
+	switch obj := o.(type) {
+	case *directfund.Objective:
+
+		for _, e := range obj.C.PreFundState().Outcome {
+			forMe := e.TotalAllocatedFor(types.AddressToDestination(fp.me))
+			for _, a := range e.Allocations {
+				if a.Amount.Cmp(forMe) != 0 {
+					slog.Warn("FairOutcomePolicy: rejecting directfund objective with unequal allocations", "objective", o.Id(), "allocations", e.Allocations, logging.WithObjectiveIdAttribute(o.Id()))
+					return false
+				}
+			}
+		}
+
+	case *virtualfund.Objective:
+		// The intermediary doesn't care about enforcing fairness
+		if obj.IsIntermediary() {
+			return true
+		}
+
+		for _, e := range obj.V.PreFundState().Outcome {
+			total := e.TotalAllocated()
+			for i, a := range e.Allocations {
+				if i == payments.PAYER_INDEX && isNotEqual(a.Amount, total) {
+					slog.Warn("FairOutcomePolicy: rejecting virtualfund objective, expected payer to start with full amount", "allocations", e.Allocations, logging.WithObjectiveIdAttribute(o.Id()))
+					return false
+				} else if i != payments.PAYER_INDEX && isNotEqual(a.Amount, big.NewInt(0)) {
+					slog.Warn("FairOutcomePolicy: rejecting virtualfund objective, expected payee to start with 0", "allocations", e.Allocations, logging.WithObjectiveIdAttribute(o.Id()))
```

Comment:
What makes the [fullAmount, 0] outcome "fair"? Is it likely anyone would want to _not_ have this policy protecting them? An alternative is to bake the check in to the protocol itself so it runs automatically (if it is that important)?

Reply:
I agree, these checks should probably just be baked into the protocol themselves.